source/language/openpulse.rst

```diff
@@ -156,12 +156,45 @@ NCO in analogy to virtual to physical register allocation.
 Frame Manipulation
 ~~~~~~~~~~~~~~~~~~
 
-The (frequency, phase) tuple of a frame can be manipulated throughout program
-by referencing ``.frequency``, and ``.phase``, with updates applied immediately
-at the current time of the frame. Operations must be appropriate for the respective type,
-``float`` for frequency and ``angle`` for phase. Again, the exact precision of these calculations
-is hardware specific. If either the frequency or phase are set to values that are invalid for
-the hardware, the compiler shall raise a compile-time error.
+The ``phase`` and ``frequency`` states of a frame can be manipulated throughout the program
+by using ``set`` and ``shift`` instructions and read using a ``get`` instruction. In particular,
+the `set_phase` and `shift_phase` instructions allow one to supply the frame and a value of type
+``angle`` representing the amount by which to set/shift the phase.
+
+.. code-block:: javascript
+
+  set_phase(angle phase, frame fr);
+  shift_frequency(angle phase, frame fr);
+
+The `get_phase` instruction allows one to supply the frame from which to retrieve the phase of
+type ``angle``.
+
+.. code-block:: javascript
+
+  get_phase(frame fr) -> angle;
+
+Analogously, the `set_frequency` and `shift_frequency` instructions allow one to supply the frame
+and a value of type ``float`` representing the amount by which to set/shift the frequency.
+
+.. code-block:: javascript
+
+  set_frequency(float freq, frame fr);
+  shift_frequency(float freq, frame fr);
+
+The `get_frequency` instruction allows one to supply the frame from which to retrieve the frequency
+of type ``float``.
+
+.. code-block:: javascript
+
+  get_frequency(frame fr) -> float;
+
+Changing the frequency or phase is an instantaneous operation. If a vendor
+is unable to support such instantaneous operations, it is expected that the
+compiler shall raise a compile-time error when encountering such frame manipulations.
+
+Moreover, the exact precision of these calculations is hardware specific. If either the frequency
+or phase are set to values that are invalid for the hardware, the compiler shall raise a
```

Comment:
I think I am interpreting this statement about frequencies to refer to a valid frequency range (as opposed to a frequency precision). What does this mean for phases?

Reply:
Good point. Upon revision, I don't think it's meaningful to discuss phase because in regards to precision we already say in the spec

> `angle name;` declares an angle with machine-architecture-specified width

Meanwhile, there is no meaningful "out of range" for phase. I've updated the text to only discuss frequency. Let me know what you think.

Reply:
The new text looks good.